crates/sui-types/src/temporary_store.rs

```diff
@@ -689,6 +688,154 @@ impl<S> TemporaryStore<S> {
     }
 }
 
+impl<S: ObjectStore> TemporaryStore<S> {
+    /// returns lists of (objects whose owner we must authenticate, objects whose owner has already been authenticated)
+    fn get_objects_to_authenticate(
+        &self,
+        sender: &SuiAddress,
+        gas: &[ObjectRef],
+    ) -> SuiResult<(HashSet<ObjectID>, HashSet<ObjectID>)> {
+        let gas_objs: HashSet<&ObjectID> = gas.iter().map(|g| &g.0).collect();
+        let mut objs_to_authenticate = HashSet::new();
+        let mut authenticated_objs = HashSet::new();
+        for (id, obj) in &self.input_objects {
+            if gas_objs.contains(id) {
+                // gas could be owned by either the sender (common case) or sponsor (if this is a sponsored tx,
+                // which we do not know inside this function).
+                // either way, no object ownership chain should be rooted in a gas object
+                // thus, consider object authenticated, but don't add it to authenticated_objs
+                continue;
+            }
+            match &obj.owner {
+                Owner::AddressOwner(a) => {
+                    assert!(sender == a, "Input object not owned by sender");
+                    authenticated_objs.insert(*id);
+                }
+                Owner::Shared { .. } => {
+                    authenticated_objs.insert(*id);
+                }
+                Owner::Immutable => {
```

Comment:
Good point! I will add a test for this one and then fix the check.

Reply:
Actually, I take it back--the code above is iterating over `input_objects` + not adding input immutable objets to the list of authenticated objects, but in your scenario I think we would have an address owned object input + immutable output.

Reply:
Ah the scenario I meant to refer to was when all the dynamic field adding happened in a previous transaction and we were now accessing dynamic fields on an immutable object in a subsequent transaction (those dynamic fields should be authenticated but if we don't add immutable objects to the set then they won't have a root to authenticate them).

Reply:
Ah got it + that makes sense. And I don't think we have a test for that, so I will add one.

Reply:
I added `crates/sui-adapter-transactional-tests/tests/dynamic_fields/read_field_from_immutable.move` to test for this, but it passes because the fields accessed do not appear in `writes` or `deleted`--the temporary store has no record of dynamic fields that only got read. Let me know if that test doesn't capture what you were thinking

However, we would definitely want to add immutable objects to the set of roots if we were doing runtime checks inside the VM instead of after-the-fact in the temporary. 

I'm not sure what the best thing to do here is--my inclination would be to keep the code as-is and comment the code heavily to explain our conversation here. My reasoning: the current approach will catch a bug where we mutate or delete an object that belongs to an immutable object (though it will show up somewhat opaquely as an authentication failure), whereas adding the immutable object to the roots will prevent us from catching this.

Reply:
Ah okay, sure, that makes sense -- the plan sounds good (heavily document but leave as-is).